### PR TITLE
Proposal for hydrodynamic added mass

### PR DESCRIPTION
**Pull request by chapulina**
* Part of https://github.com/gazebosim/gz-sim/issues/1462

I'm looking for early feedback on the proposal. I'm planning to circulate this with the [ROS Maritime Working Group](https://discourse.ros.org/t/new-working-group-proposal-maritime-robotics/23832) for feedback after the initial review.

Rendered preview:

http://sdformat.org/tutorials?tut=added_mass_proposal&branch=chapulina/added_mass

### CODE REVIEW

**Comment by scpeters** [comment]
For people unfamiliar with hydrodynamics, the term "added mass" is not very clear. Can you use the phrase "hydrodynamic added mass" in titles and when introducing the concept to help with clarifying this?

**Comment by hamilton8415** [comment]
Thanks Steve, I will work through this and add some clarifying details as you suggest.  Trick will be for me to keep it short, suggestions on how general an audience to target are welcome.

**Comment by scpeters** [comment]
> Thanks Steve, I will work through this and add some clarifying details as you suggest. Trick will be for me to keep it short, suggestions on how general an audience to target are welcome.

adding the word "hydrodynamic" is probably enough for me at first glance, but I'll add additional questions as they occur to me

File: added_mass/proposal.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    {\bf x}^T 
+    =
+    \begin{bmatrix}
+      x           & y           & z           & p           &  q         & r 
+    \end{bmatrix}
+$$
+
+where,
+
+* $x$ : Position in X axis
```
Comment (chapulina) [inline added_mass/proposal.md:54]: @hamilton8415 , I know it's a bit odd, but the previous style (`\\(..\\)`) is the only one I got to render correctly for inline code with MathJax. It [explicitly disallows](https://docs.mathjax.org/en/latest/input/tex/delimiters.html) `$..$` inline by default. You can see this is not being displayed correctly on the preview link. 

```suggestion
* \\(x\\): Position in X axis
```

File: added_mass/proposal.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+/// of the mass matrix are specified (see this class’s documentation for
+/// details). The pose object specifies the pose X_FBi of the inertial
+/// frame Bi in the frame F of this inertial object
+/// (see class’s documentation). The added mass matrix is also expressed
+/// in frame Bi.
+/// \param[in] _massMatrix Mass and inertia matrix.
+/// \param[in] _pose Pose of center of mass reference frame.
+public: Inertial(const MassMatrix3<T> &_massMatrix,
+                 const Pose3<T> &_pose,
+                 const Matrix6<T> &_addedMass)
```
Comment (scpeters) [inline added_mass/proposal.md:226]: I posted an alternative to `Matrix6` in https://github.com/gazebosim/gz-math/pull/442#issuecomment-1159427804

File: added_mass/proposal.md (hunk 1 above)
Comment (chapulina) [reply added_mass/proposal.md]: I went ahead and fixed all instances of this

File: added_mass/proposal.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+The [`//link/inertial`](http://sdformat.org/spec?ver=1.9&elem=link#link_inertial) term
+in SDF currenly accepts elements to describe the m matrix (mass and moments of inertia),
+as well as a center of mass pose. This proposal suggests exposing each of the 21 elements
+for the mass matrix in addition to that. All added mass elements are optional and default
+to zero if unset. This preserves the behaviour for links that don't have those terms.
+
+### XML spec
+
+A new `<added_mass>` element will be added under `//link/inertial/`. It will contain each of the
+21 matrix elements started by `n` and followed by its index.
```
[the remark this answers is not in the document]
Comment (chapulina) [reply added_mass/proposal.md]: I like the change of elements, it also has the advantage of mirroring the notation in `inertia`.

[the remark this answers is not in the document]
Comment (chapulina) [reply added_mass/proposal.md]: For the `hydrodynamic` part, one concern is that these terms also apply to other objects submerged in fluids other than water, like air. Not sure if something like `fluid_added_mass` makes enough sense?

[the remark this answers is not in the document]
Comment (hamilton8415) [reply added_mass/proposal.md]: All of this looks fine for our purposes here.  
- For me, \mu is the most natural and general, but the letter notation is fine too, if a bit non-standard.  But mirroring the inertia tensor sounds good.
- In the fluid dynamics world, air is most certainly understood to be a fluid so that distinction doesn't keep me up at night at all.  And you are right, in the case of a balloon in air, the added mass is a major contribution to inertia...   
-  Similarly, in my experience "added mass" is a very common noun so I would hesitate to insert the world fluid in the middle.  Actually, I would eliminate the inclusion of "fluid" entirely and simply refer to these quantities as "added mass", the problem is the fluid is causing the added inertia, so people usually refer to the bodies added mass, not the fluid.  Ugh, semantics.

[the remark this answers is not in the document]
Comment (chapulina) [reply added_mass/proposal.md]: I adopted the term "fluid added mass" throughout the proposal in [0c24fbc](https://github.com/gazebosim/sdf_tutorials/pull/76/commits/0c24fbcf822d1ce879129ba9ed6f96a0d3f79713). I believe all items in this comment have been addressed.

**Comment by chapulina** [comment]
I pushed some more paragraphs in [5c36c87](https://github.com/gazebosim/sdf_tutorials/pull/76/commits/5c36c87bd2e971b1df8596c241b2df1335cc6b12) explaining the motivation for storing the added mass terms separately from the remaining inertial terms, even though they will be combined when fed to the physics engine.

I also updated the target version to Garden, since any changes to `math::Inertial` will break ABI. This means we have 1 month to get at least the math changes in.

File: added_mass/proposal.md (hunk 2 above)
Comment (chapulina) [reply added_mass/proposal.md:232]: I'll take a stab at storing the elements as a `std::array<double, 21>` instead of a `Matrix6d`, mimicking the idea from how `MassMatrix3`stores the moments in vectors instead of a matrix. My main concern is that this may result in a very clunky API.

File: added_mass/proposal.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+* \\(qq\\) : added mass moment about the Y axis due to angular acceleration about the X direction
+* \\(qr\\) : added mass moment about the Y axis due to angular acceleration about the X direction, and vice-versa
+* \\(rr\\) : added mass moment about the Z axis due to angular acceleration about the Z direction
+
+## Proposed implementation
+
+While many physics engines may not have direct support for added mass, they often
+support directly setting each element of the resulting \\(\bf{M} + \bf{\mu}\\)
+matrix that's multiplied by acceleration. That's the case for
+[DART](https://dartsim.github.io/dart/main/d6/d91/classdart_1_1dynamics_1_1Inertia.html#aa4661f9950d5958efa4c8609d42aedf7),
```
Comment (scpeters) [inline added_mass/proposal.md:151]: I was just looking more closely at DART's inertia code, and it looks like it will print warnings if a spatial inertia matrix has nonzero values in off-diagonal parts of the translational 3x3 block-diagonal matrix (`xy`, `xz`, `yz`) or the diagonal parts of the block-off-diagonal 3x3 matrix (`xp`, `yq`, `zr`). For rigid bodies, these terms are all `0`, so I think DART will complain a bit about using them for fluid added mass, though it might still work despite the complaints, depending on what assumptions they make when inverting the spatial inertia matrix

Comment (chapulina) [reply added_mass/proposal.md:151]: Yeah that's a concern that @hamilton8415 and I have had from the beginning. Hopefully it's just a matter of suppressing those warnings. It would be nice to have a prototype implementation out soon so we can have an idea of what breaks and what doesn't.

File: added_mass/proposal.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+public: bool SetFluidAddedMass(const Matrix6<T> &_m)
+
+/// \brief Get the added mass matrix.
+/// \return The added mass matrix.
+public: const Matrix6<T> &FluidAddedMass() const
+
+/// \brief Spatial mass matrix, which is the resulting 6x6 matrix including
+/// added mass.
+/// \return The spatial mass matrix.
+public: const Matrix6<T> &SpatialMatrix() const
```
Comment (scpeters) [inline added_mass/proposal.md:247]: thanks for your patience; this draft of the proposal looks good so far. I think we just need to clarify the order of variables used in this spatial matrix. The matrix earlier in this document uses `[x y z p q r]` ordering, but I believe dartsim and drake have the rotational variables first `[p q r x y z]`. We don't need to arbitrary parameter ordering but just document the order

Comment (scpeters) [reply added_mass/proposal.md:247]: I expect that variable ordering convention is used in dartsim and drake because it is the one used by Featherstone:

* https://homes.cs.washington.edu/~todorov/courses/amath533/FeatherstoneSlides.pdf

File: added_mass/proposal.md [hunk 6]
```diff
@@ -121,27 +121,27 @@ $$
...
+* \\(yr\\) : added mass in the Y axis due to angular acceleration about the X axis, and vice-versa
+* \\(zz\\) : added mass in the Z axis due to linear acceleration in the Z axis
+* \\(zp\\) : added mass in the Z axis due to angular acceleration about the X axis, and vice-versa
+* \\(zq\\) : added mass in the Z axis due to angular acceleration about the X axis, and vice-versa
+* \\(zr\\) : added mass in the Z axis due to angular acceleration about the X axis, and vice-versa
+* \\(pp\\) : added mass moment about the X axis due to angular acceleration about the X axis
+* \\(pq\\) : added mass moment about the X axis due to angular acceleration about the X axis, and vice-versa
+* \\(pr\\) : added mass moment about the X axis due to angular acceleration about the X axis, and vice-versa
+* \\(qq\\) : added mass moment about the Y axis due to angular acceleration about the X axis
+* \\(qr\\) : added mass moment about the Y axis due to angular acceleration about the X axis, and vice-versa
```
Comment (scpeters) [inline added_mass/proposal.md:143]: all instances of "due to angular acceleration about the _ axis" on this line and above refer to the `X` axis, which is incorrect in most cases. the following line documenting `rr` is correct

File: added_mass/proposal.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+      xr & yr & zr & pr & qr & rr
+    \end{bmatrix}
+$$
+
+where
+
+* \\(xx\\) : added mass in the X axis due to linear acceleration in the X axis
+* \\(xy\\) : added mass in the X axis due to linear acceleration in the Y axis, and vice-versa
+* \\(xz\\) : added mass in the X axis due to linear acceleration in the Z axis, and vice-versa
+* \\(xp\\) : added mass in the X axis due to angular acceleration about the X axis, and vice-versa
```
Comment (arjo129) [inline added_mass/proposal.md:127]: Would it be worth specifying units?

**Review by arjo129 — COMMENTED** [review]
Is it worth describing the units here.

File: added_mass/proposal.md (hunk 7 above)
Comment (chapulina) [reply added_mass/proposal.md:131]: I added the units, but I'd like to confirm with @hamilton8415 or @andermi if they're correct.

**Comment by chapulina** [comment]
While working on the implementation I realized that since the added mass coefficients are fluid-dependent, it would be convenient to expose the fluid density as a separate parameter. This way, it's trivial to adapt an SDF file for different fluids. 

I made the changes for that in 60866d47aab1a84b78edd83c301db865e217a11f, and will start working on the implementation. I'd appreciate some reviews, since I haven't seen the added mass matrix broken apart like this anywhere in the literature. I may be doing something that may not hold for every use case, or that will make it difficult to export values from other software directly into an SDF file.

CC @andermi

**Comment by andermi** [comment]
Yes, some bodies will have empirical added mass matrices from CFD or pool tests, and some easier shapes will have closed-form dependent on fluid density. Perhaps, we could support an explicit matrix and a formulaic one?

File: added_mass/proposal.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+in SDF currenly accepts elements to describe the m matrix (mass and moments of inertia),
+as well as a center of mass pose. This proposal suggests exposing each of the 21 elements
+for the mass matrix in addition to that. All added mass elements are optional and default
+to zero if unset. This preserves the behaviour for links that don't have those terms.
+
+A new `<fluid_added_mass>` element will be added under `//link/inertial/`.
+It will contain each of the 21 matrix elements, and the fluid density.
+
+The fluid density's unit is \\(kg/m^3\\), and the unit of each matrix element is \\(m^3\\),
+so that the unit of the final matrix is \\(kg\\).
```
[the remark this answers is not in the document]
Comment (chapulina) [reply added_mass/proposal.md]: Nice catch, @JoanAguilar also noticed it. I hope I got them right now.

**Comment by chapulina** [comment]
> expose the fluid density as a separate parameter

After talking to @andermi today we figured that wasn't a very good idea, so I removed the separate parameter.

File: added_mass/proposal.md (hunk 5 above)
Comment (chapulina) [reply added_mass/proposal.md]: Updated

File: added_mass/proposal.md (hunk 4 above)
Comment (chapulina) [reply added_mass/proposal.md:155]: From my initial testing, all we need to do is suppress those warnings, see 

* https://github.com/dartsim/dart/pull/1672

I'll check with @hamilton8415 in the coming days if the resulting behavior is what's expected though.

**Comment by mjcarroll** [comment]
After landing the implementation PR, we are going to give this one more review to confirm that the implementation and documentation match.

File: added_mass/proposal.md [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+The proposal includes the following sections:
+
+* [Motivation](#motivation): An explanation of the background and rationale behind the proposal
+* [Proposed implementation](#proposed-implementation): Terms to be added to the SDFormat specification and C++ API
+
+## Motivation
+
+Newton's second law describes how forces affect a body's motion in an inertial frame. It can be written as:
+
+$$ (\bf{M} + \bf{\mu})   \ddot{\bf x} = \sum{F({\bf x}, t)} $$
```
Comment (JoanAguilar) [inline added_mass/proposal.md:39]: Note that this equation is only valid for translation; for the rotation of a rigid body, this becomes [Euler's equations](https://en.wikipedia.org/wiki/Euler%27s_equations_(rigid_body_dynamics)); and in the presence of cross-terms in the spatial inertia matrix, this becomes equation (2.90) from [1].
